### DISCUSSION TITLE
Communication between Linux(Master) and Linux(Slave) on Desktop Core 0 and Core 1 homogeneous processor

### DISCUSSION THREAD

**Discussion by viswaharsha** [OpenAMP/open-amp#289] [Q&A]
Hi,

I am using my desktop core i5 quad core processor.
I have two terminals of same linux OS running on core0(terminal 1) and core1(terminal 2) using taskset.
Can they communicate between each other as Master and Slave and perform remote procedure calls between each other available at apps/examples/rpc_demo/demod.c for master(terminal 1) and apps/examples/rpc_demo/demo.c for slave(terminal 2)?

**Comment by arnopo** [comment]
Do you try to run exemples on your Linux PC, implementation seems close to your need (except the use of taskset for CPU affinity)?
https://github.com/OpenAMP/open-amp#example-to-compile-openamp-for-communication-between-linux-processes
To understand communication between the processes, you can have a look here:
https://github.com/OpenAMP/open-amp/blob/master/apps/system/linux/machine/generic/platform_info.c

**Reply by viswaharsha** [reply]
Yes, I have followed these steps to compile and build the library. I have tried to run remote procedure calls demo app available at (https://github.com/OpenAMP/open-amp/tree/master/apps/examples/rpc_demo)
The endpoints between the terminals are connected and the slave program is performing I/O and file operations locally. But the purpose is, the slave program has to perform these operations in the terminal which is running the master program.

**Reply by arnopo** [reply]
> Yes, I have followed these steps to compile and build the library. I have tried to run remote procedure calls demo app available at (https://github.com/OpenAMP/open-amp/tree/master/apps/examples/rpc_demo)

I suppose that you have adapted the demo for this, right? because it has been designed to run on Xilinx platform(cf [xil_printf](https://github.com/OpenAMP/open-amp/blob/c3f12991991d187d47637d88e8a66076f23d0c8c/apps/examples/rpc_demo/rpc_demo.c#L28) )

> The endpoints between the terminals are connected and the slave program is performing I/O and file operations locally. But the purpose is, the slave program has to perform these operations in the terminal which is running the master program.

Not sure to follow you...you mentioned in you first comment that you have 2 terminals ...but here you speak about only one terminal to run both programs.
Please clarify your setup:
- patch applied on existing demo.
- build commands
- command used to run program in terminal(s)

**Reply by viswaharsha** [reply]
Yes, I have modified some code to make it work on non-xilinx platform(modified xil_printf to printf)

**Build Command:**

`cmake /home/viswa/Desktop/open-amp -DCMAKE_INCLUDE_PATH=/home/viswa/Desktop/libmetal/build/usr/local/include -DCMAKE_LIBRARY_PATH=/home/viswa/Desktop/libmetal/build/usr/local/lib -DWITH_APPS=ON`

`sudo make VERBOSE=1 DESTDIR=$(pwd) install` current directory is `open-amp/build-openamp`

**Run Command:**

**Terminal 1(Master) running on core 1(CPU affinity set using **tasket**):**
`sudo LD_LIBRARY_PATH=/home/viswa/Desktop/open-amp/build-openamp/usr/local/lib: /home/viswa/Desktop/open-amp/build-openamp/apps/examples/rpc_demo/rpc_demod-shared`

**Terminal 2(Slave) running on core 2(CPU affinity set using **taskset**):**
`sudo LD_LIBRARY_PATH=/home/viswa/Desktop/open-amp/build-openamp/usr/local/lib: /home/viswa/Desktop/open-amp/build-openamp/apps/examples/rpc_demo/rpc_demo-shared 1`

**These terminals run on different directories on same operating system.**

**Reply by arnopo** [reply]
Thanks for the detail.

I did a short test on my side using one terminal

Patches on code:
```

-------------------- apps/examples/rpc_demo/CMakeLists.txt --------------------
index 5d5655de4a0e..ba14b142cb4c 100644
@@ -18,7 +18,7 @@ collector_list (_deps PROJECT_LIB_DEPS)
 set (OPENAMP_LIB open_amp)
 
 if (${PROJECT_SYSTEM} STREQUAL "linux")
-  set (app_list rpc_demod)
+  set (app_list rpc_demod rpc_demo)
 else (${PROJECT_SYSTEM})
   set (app_list rpc_demo)
 endif (${PROJECT_SYSTEM} STREQUAL "linux")

---------------------- apps/examples/rpc_demo/rpc_demo.c ----------------------
index a11cc8a5213c..646198b15cfb 100644
@@ -25,7 +25,7 @@
 #define REDEF_O_APPEND  0002000
 #define REDEF_O_ACCMODE 0000003
 
-#define LPRINTF(format, ...) xil_printf(format, ##__VA_ARGS__)
+#define LPRINTF(format, ...) printf(format, ##__VA_ARGS__)
 //#define LPRINTF(format, ...)
 #define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)

```
script to build and run ( on the same terminal)
```
 #!/bin/bash
 
dirs_to_check=$PWD
echo "################### build libmetal #####################"
cd libmetal
rm -r build
cmake . -Bbuild || exit 1
cd build
make || exit 1

cd $dirs_to_check

echo "################### build openamp #####################"
cd open-amp
rm -r build
cmake . -DWITH_APPS=on -DWITH_PROXY=on -Bbuild -DCMAKE_INCLUDE_PATH="../libmetal/build/lib/include" -DCMAKE_LIBRARY_PATH="../libmetal/build/lib" || exit 1
cd build
make install DESTDIR=./bin || exit 1

cd $dirs_to_check
cd open-amp
cd build

echo "################### run rpc demo #####################"
LD_LIBRARY_PATH=./bin/usr/local/lib:../../libmetal/build/lib/  ./bin/usr/local/bin/rpc_demod-shared >/dev/null&
sleep 1
LD_LIBRARY_PATH=./bin/usr/local/lib:../../libmetal/build/lib/  ./bin/usr/local/bin/rpc_demo-shared 1
```

I tested using two separate terminals: test OK

The main difference i can see with your build command is the  **`-DWITH_PROXY=on`**
Then i don't use neither taskset nor sudo rights.

Hope that this would help

**Reply by viswaharsha** [reply]
**Shared below is the screenshot of the terminal 1(left image) running masterprogram(rpc_demod-shared) and terminal 2(right image) running slave program(rpc_demo-shared)**

![image](https://user-images.githubusercontent.com/54477309/118709933-3d8a7f00-b83b-11eb-8816-42cdbcb52b50.png)

The expectation is, the file operation(open, read, write and close) should be performed on the terminal running master(rpc_demod-shared) and **file has to be created on the directory named 'master'** but the file operations are performed in the terminal running slave(rpc_demo-shared) and the **file is created on directory named 'slave'** as **remote.file**. 

Also the input(name, age, value of pi) has to be fetched from the user using the terminal running master(rpc_demod-shared) and displayed in terminal running slave(rpc_demo-shared). But the input(name, age, value of pi) is fetched from the user in the terminal running slave(rpc_demo-shared) and is displayed in the terminal running slave(rpc_demo-shared).

**Reply by arnopo** [reply]
Here your point is no more the IPC communication but to adapt the demo to redirect file operations over rpc, right?
I don't think that it is a good idee as the libmetal uses file operations for the shared memory management.
So you will have to addapt the libmetal and/or the demo +  [rpmsg_retarget.c](https://github.com/OpenAMP/open-amp/blob/master/lib/proxy/rpmsg_retarget.c) file to your need.

Then you will probably also have to fix the fact that two libraries implement the file operation. In this case (if I remember correctly) the linker chooses the first one it finds. One solution for this is to rename and export _open, _close defined in  [rpmsg_retarget.c](https://github.com/OpenAMP/open-amp/blob/master/lib/proxy/rpmsg_retarget.c) and to redefine the file operations in the demo code.

**Reply by viswaharsha** [reply]
I have applied some changes to the following files to provide the output that I have attached below
+ rpc_demod.c
+ rpc_demo.c 
+ rpmsg_retarget.c

![image](https://user-images.githubusercontent.com/54477309/119310770-df6ff880-bc8d-11eb-984f-bdc171d9c12c.png)

So can you kindly check and let me know the procedure to share the patch and test the code for contribution

**Reply by arnopo** [reply]
To contribute to the OpenAMP library please follow the process described in the [README.md](https://github.com/OpenAMP/open-amp#how-to-contribute)

**Reply by Raveem13** [reply]
> **Shared below is the screenshot of the terminal 1(left image) running masterprogram(rpc_demod-shared) and terminal 2(right image) running slave program(rpc_demo-shared)**
> 
> ![image](https://user-images.githubusercontent.com/54477309/118709933-3d8a7f00-b83b-11eb-8816-42cdbcb52b50.png)
> 
> The expectation is, the file operation(open, read, write and close) should be performed on the terminal running master(rpc_demod-shared) and **file has to be created on the directory named 'master'** but the file operations are performed in the terminal running slave(rpc_demo-shared) and the **file is created on directory named 'slave'** as **remote.file**.
> 
> Also the input(name, age, value of pi) has to be fetched from the user using the terminal running master(rpc_demod-shared) and displayed in terminal running slave(rpc_demo-shared). But the input(name, age, value of pi) is fetched from the user in the terminal running slave(rpc_demo-shared) and is displayed in the terminal running slave(rpc_demo-shared).

@viswaharsha How did you assigned cpus to terminals?

**Comment by arnopo** [comment]
@edmooring 
Any reason to not adapt the demo for Linux to Linux communication ( as done for other examples)?
Could we replace xil_printf by a generic printf?

**Reply by edmooring** [reply]
@arnopo 
I can't think of any reason not to do so. xil_printf() takes the same arguments as the generic printf(). In theory, it might be a little bit smaller code, but I don't know if anyone has ever measured it.
